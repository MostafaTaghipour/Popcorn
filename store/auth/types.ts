import { Action } from "redux";
import { ApiAction, RequestState, PostRequestState } from "@app/types/action";

export interface AuthRequestBody {
  username: string;
  password: string;
}

export interface AuthResponseBody {
  token: string;
}

// auth state model
export interface AuthState {
  /**
   * authenticate request state
   *
   * @type {PostRequestState}
   * @memberof AuthState
   */
  authentication: PostRequestState;
  /**
   * user token
   *
   * @type {string}
   * @memberof AuthState
   */
  token?: string;
}

// actions
export enum AuthActionTypes {
  AUTHENTICATION = "AUTHENTICATION",
  SIGN_OUT = "SIGN_OUT",
}

export interface AuthenticationAction
  extends ApiAction<AuthActionTypes.AUTHENTICATION, AuthResponseBody> {}

export interface SignOutAction extends Action<AuthActionTypes.SIGN_OUT> {}

export type AuthActions = AuthenticationAction | SignOutAction;
