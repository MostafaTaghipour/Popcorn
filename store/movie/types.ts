import { Action } from "redux";
import {
  ApiAction,
  RequestState,
  PaginationRequestState,
  ApiPaginationAction,
} from "@app/types/action";

export interface ListResult<T> {
  count: number;
  next?: string;
  previous?: string;
  results: T[];
}

export interface Movie {
  id: number;
  title: string;
  date_of_release: string;
  rating: number;
  tags: string[];
  director: string;
  poster?: string;
}

export interface Category {
  id: number;
  name: string;
}

// movie state model
export interface MovieState {
  currentCategoryMovies: PaginationRequestState<Movie[]>;
  lastSeenCategory?: string;
  searchResultMovies: PaginationRequestState<Movie[]>;
  categories: RequestState<Category[]>;
}

// actions
export enum MovieActionTypes {
  SEARCH_MOVIES = "SEARCH_MOVIES",
  FETCH_CATEGORY_MOVIES = "FETCH_CATEGORY_MOVIES",
  FETCH_CATEGORIES = "FETCH_CATEGORIES",
  CLEAR_SEARCH_RESULTS = "CLEAR_SEARCH_RESULTS",
}

export interface FetchCategoryMoviesAction
  extends ApiPaginationAction<MovieActionTypes.FETCH_CATEGORY_MOVIES, Movie[]> {
  payload: string;
}

export interface SearchMoviesAction
  extends ApiPaginationAction<MovieActionTypes.SEARCH_MOVIES, Movie[]> {}

export interface FetchCategoriesAction
  extends ApiAction<MovieActionTypes.FETCH_CATEGORIES, Category[]> {}

export interface ClearSearchResultAction
  extends Action<MovieActionTypes.CLEAR_SEARCH_RESULTS> {}

export type MovieActions =
  | FetchCategoryMoviesAction
  | SearchMoviesAction
  | FetchCategoriesAction
  | ClearSearchResultAction;
