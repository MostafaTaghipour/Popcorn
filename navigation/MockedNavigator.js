import React from "react";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";
import { Provider as ReduxProvider } from "react-redux";
import { AuthProvider } from "../contexts/AuthContext";

const Stack = createStackNavigator();

/**
 * Provide a MockedNavigator to wrap our screens
 * 
 * Access navigation in screens 
 * 
 * @param {*} param0 
 */
export const MockedNavigator = ({ component, params = {}, store }) => {
  return (
    <ReduxProvider store={store}>
      <AuthProvider>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen
              name="MockedScreen"
              component={component}
              initialParams={params}
            />
          </Stack.Navigator>
        </NavigationContainer>
      </AuthProvider>
    </ReduxProvider>
  );
};
