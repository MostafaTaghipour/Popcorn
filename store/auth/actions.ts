import {
  AuthActionTypes,
  AuthenticationAction,
  AuthRequestBody,
  SignOutAction,
} from "./types";
import { AsyncActionStatus } from "@app/types/action";
import HttpStatusCode from "@app/types/net";
import { showToastAction } from "../ui/actions";
import LocaleManager from "@app/helpers/LocaleManager";
import AuthApi from "@app/data/remote/AuthApi";
import { AxiosError } from "axios";
import mocker from "@app/net/mocker";

/**
 * authenticate user action
 *
 * @param {string} username
 * @param {string} password
 * @return {*}
 */
export const authAsyncAction = (username: string, password: string) => {
  return async (dispatch: any) => {
    var action: AuthenticationAction = {
      type: AuthActionTypes.AUTHENTICATION,
      status: AsyncActionStatus.REQUEST,
    };

    dispatch(action);

    try {
      const body: AuthRequestBody = { username, password };
      const response = await AuthApi.authenticate(body);

      action = {
        ...action,
        status: AsyncActionStatus.SUCCESS,
        data: response.data,
      };
      dispatch(action);
    } catch (error) {
      action = {
        ...action,
        status: AsyncActionStatus.FAILURE,
        error: error,
      };
      dispatch(action);

      if ((error as AxiosError)?.response?.status == HttpStatusCode.BadRequest)
        dispatch(
          showToastAction({
            text: LocaleManager.t("auth.wrong_credentials"),
            type: "danger",
          })
        );
    }
  };
};

export const signOutAction = (): SignOutAction => {
  return {
    type: AuthActionTypes.SIGN_OUT,
  };
};
