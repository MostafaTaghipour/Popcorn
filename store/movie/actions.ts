import {
  FetchCategoriesAction,
  MovieActionTypes,
  Movie,
  FetchCategoryMoviesAction,
  SearchMoviesAction,
  ClearSearchResultAction,
} from "./types";
import { AsyncActionStatus, PaginationRequestState } from "@app/types/action";
import MovieApi from "@app/data/remote/MovieApi";
import { store } from "..";
import Configs from "@app/constants/Configs";
/**
 * fetch movie categories action
 *
 * @return {*}
 */
export const fetchCategoriesAsyncAction = () => {
  return async (dispatch: any) => {
    var action: FetchCategoriesAction = {
      type: MovieActionTypes.FETCH_CATEGORIES,
      status: AsyncActionStatus.REQUEST,
    };

    dispatch(action);

    try {
      const response = await MovieApi.categories();

      action = {
        ...action,
        status: AsyncActionStatus.SUCCESS,
        data: response.data.results,
      };
      dispatch(action);
    } catch (error) {
      action = {
        ...action,
        status: AsyncActionStatus.FAILURE,
        error: error,
      };
      dispatch(action);
    }
  };
};

/**
 * fetch specific movie category movie action
 *
 * @return {*}
 */
export const fetchCategoryMovieAsyncAction = (
  category: string,
  pageNumber = Configs.PAGINATION_FIRST_PAGE_NUMBER,
  refreshing = false
) => {
  return async (dispatch: any) => {
    const currentState: PaginationRequestState<Movie[]> = store.getState().movie
      .currentCategoryMovies;

    //check if all data loaded
    if (currentState.data!.length >= currentState.total && !refreshing) return;

    //check if already loading
    if (currentState.loading || currentState.refreshing) return;

    //check is refreshing or loading more
    const page = refreshing ? Configs.PAGINATION_FIRST_PAGE_NUMBER : pageNumber;

    var action: FetchCategoryMoviesAction = {
      type: MovieActionTypes.FETCH_CATEGORY_MOVIES,
      status: AsyncActionStatus.REQUEST,
      payload: category,
      refreshing: refreshing,
      page: page,
    };
    dispatch(action);

    

    try {
      const limit = Configs.PAGINATION_PAGE_SIZE;
      const offset = page * limit;
      const response = await MovieApi.categoryMovies(category, limit, offset);

      action = {
        ...action,
        status: AsyncActionStatus.SUCCESS,
        data: response.data.results,
        total: response.data.count,
      };
      dispatch(action);
    } catch (error) {
      action = {
        ...action,
        status: AsyncActionStatus.FAILURE,
        error: error,
      };
      dispatch(action);
    }
  };
};

/**
 * search movie action
 *
 * @return {*}
 */
export const searchMovieAsyncAction = (
  query: string,
  pageNumber = Configs.PAGINATION_FIRST_PAGE_NUMBER,
  refreshing = false
) => {
  return async (dispatch: any) => {
    const currentState: PaginationRequestState<Movie[]> = store.getState().movie
      .searchResultMovies;

    //check if all data loaded
    if (currentState.data!.length >= currentState.total && !refreshing) return;

    //check if already loading
    if (currentState.loading || currentState.refreshing) return;

    //check is refreshing or loading more
    const page = refreshing ? Configs.PAGINATION_FIRST_PAGE_NUMBER : pageNumber;

    var action: SearchMoviesAction = {
      type: MovieActionTypes.SEARCH_MOVIES,
      status: AsyncActionStatus.REQUEST,
      refreshing: refreshing,
      page: page,
    };
    dispatch(action);

    try {
      const limit = Configs.PAGINATION_PAGE_SIZE;
      const offset = page * limit;
      const response = await MovieApi.search(query, limit, offset);

      action = {
        ...action,
        status: AsyncActionStatus.SUCCESS,
        data: response.data.results,
        total: response.data.count,
      };
      dispatch(action);
    } catch (error) {
      action = {
        ...action,
        status: AsyncActionStatus.FAILURE,
        error: error,
      };
      dispatch(action);
    }
  };
};

/**
 * clear old search results
 *
 * @return {*}  {ClearSearchResultAction}
 */
export const clearSearchResultsAction = (): ClearSearchResultAction => {
  return {
    type: MovieActionTypes.CLEAR_SEARCH_RESULTS,
  };
};
