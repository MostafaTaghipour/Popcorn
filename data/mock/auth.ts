import { Movie, Category } from "@app/store/movie/types";
import { AuthResponseBody } from "@app/store/auth/types";

export const mockAuthResponse: AuthResponseBody = {
  token: "TOKEN",
};
