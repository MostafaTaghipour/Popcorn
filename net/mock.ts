import { ListResult } from "./../store/movie/types";
/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */
import mocker from "./mocker";
import { mockCategories, mockMovies } from "@app/data/mock/movie";
import { mockAuthResponse } from "@app/data/mock/auth";

/**
 * we can define our mock data for each request
 *
 */
export const provideMockData = () => {
  mocker.addMock("user/auth-token", {
    data: (req: any) => {
      if ((req.data.username = "mostafa" && req.data.password == "123"))
        return mockAuthResponse;

      let err = Error("wrong credentials");

      //@ts-ignore
      err.code = 400;

      return err;
    },
    method: "post",
  });
  mocker.addMock("category", {
    data: { count: mockCategories.length, results: mockCategories },
    method: "get",
  });
  mocker.addMock("tags", {
    data: { count: mockMovies.length, results: mockMovies },
    method: "get",
  });
  mocker.addMock("search", {
    data: { count: mockMovies.length, results: mockMovies },
    method: "get",
  });
};
