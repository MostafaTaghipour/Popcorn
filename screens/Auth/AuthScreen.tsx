import {
  StyleSheet,
  KeyboardAvoidingView,
  Platform,
  View,
  Keyboard,
} from "react-native";
import React, { useContext, useRef, useState, useEffect } from "react";
import {
  Title,
  TextInput,
  ActivityIndicator,
  Button,
  Text,
} from "react-native-paper";
import { LocalizationContext } from "@app/contexts/LocalizationContext";
import { Feather } from "@expo/vector-icons";
import { useNavigation } from "@react-navigation/native";
import { useDispatch } from "react-redux";
import { authAsyncAction } from "@app/store/auth/actions";
import { ThemeContext } from "@app/contexts/ThemeContext";
import { useReduxSelector } from "@app/store";
import { ConnectivityContext } from "@app/contexts/ConnectivityContext";
import { showToastAction } from "@app/store/ui/actions";
/**
 * AuthScreen for user authentication
 *
 * @export
 * @return {*}
 */
export default function AuthScreen() {
  const { t } = useContext(LocalizationContext);
  const { theme, isDark } = useContext(ThemeContext);
  const { isConnected } = useContext(ConnectivityContext);
  const state = useReduxSelector((state) => state.auth.authentication);
  const dispatch = useDispatch();
  const navigation = useNavigation();
  const userNameRef = useRef(null);
  const [userName, setUserName] = useState("");
  const [password, setPassword] = useState("");
  const [showPassword, setShowPassword] = useState(false);

  // focus on user name filed on start
  useEffect(() => {
    setTimeout(() => {
      //@ts-ignore
      userNameRef.current?.focus();
    });
  }, []);

  // close screen when authentication done
  useEffect(() => {
    if (state.done) close();
  }, [state.done]);

  const disableButton: boolean = React.useMemo(() => {
    return !userName || !password;
  }, [userName, password]);

  const close = () => {
    navigation.goBack();
  };
  /**
   * login action if connect to internet
   *
   */
  const login = () => {
    if (isConnected) dispatch(authAsyncAction(userName, password));
    else
      dispatch(
        showToastAction({
          text: t("error.no_internet"),
          type: "danger",
        })
      );
    Keyboard.dismiss();
  };

  return (
    <KeyboardAvoidingView
      style={styles.flex_1}
      behavior={Platform.select({ android: "height", ios: "padding" })}
    >
      <View style={styles.topSection}>
        <Title>{t("auth.title")}</Title>
        <Feather
          style={styles.closeButton}
          name="x"
          size={30}
          onPress={close}
          color={isDark ? "white" : "black"}
        />
      </View>
      <TextInput
        disabled={state.inProgress}
        mode="outlined"
        testID="UserName"
        label={t("auth.user_name")}
        ref={userNameRef}
        onChangeText={setUserName}
        style={styles.input}
      />

      <View style={styles.input}>
        <TextInput
          testID="Password"
          disabled={state.inProgress}
          mode="outlined"
          label={t("auth.password")}
          onChangeText={setPassword}
          secureTextEntry={!showPassword}
        />

        <Feather
          name={showPassword ? "eye" : "eye-off"}
          size={24}
          style={styles.eyeButton}
          onPress={() => setShowPassword(!showPassword)}
          color={theme.colors.primary}
        />
      </View>
      <Button
        disabled={disableButton || state.inProgress}
        mode="contained"
        style={styles.button}
        onPress={login}
      >
        <Text
          testID="Submit"
          accessibilityHint={disableButton.toString()}
        >
          {t("auth.login")}
        </Text>
        {state.inProgress && (
          <ActivityIndicator
            animating={true}
            style={styles.buttonIndicator}
            color={theme.colors.primary}
          />
        )}
      </Button>
    </KeyboardAvoidingView>
  );
}

const styles = StyleSheet.create({
  flex_1: {
    flex: 1,
  },
  topSection: {
    height: 60,
    justifyContent: "center",
    alignItems: "center",
  },
  input: {
    marginVertical: 8,
    marginHorizontal: 16,
  },
  button: {
    width: 200,
    marginTop: 32,
    alignSelf: "center",
    height: 50,
    justifyContent: "center",
    alignItems: "center",
  },
  closeButton: {
    position: "absolute",
    left: 16,
  },
  eyeButton: {
    position: "absolute",
    right: 16,
    top: 24,
  },
  buttonIndicator: {
    marginStart: 16,
    width: 20,
    height: 20,
  },
});
