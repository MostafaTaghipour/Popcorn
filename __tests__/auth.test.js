import React from "react";
import AuthScreen from "../screens/Auth/AuthScreen";
import { MockedNavigator } from "../navigation/MockedNavigator";
import { render, fireEvent } from "react-native-testing-library";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import mocker from "../net/mocker";
import { AuthActionTypes } from "../store/auth/types";
import { authAsyncAction } from "../store/auth/actions";
import { mockAuthResponse } from "../data/mock/auth";
import { AsyncActionStatus } from "../types/action";

/**
 * test AuthScreen rendering
 */
describe("Auth render", () => {
  let initialState;
  let store;

  /**
   * initialize mock redux store
   */
  beforeAll(() => {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);

    // Initialize mockstore
    initialState = {
      auth: {
        authentication: {},
        token: undefined,
      },
    };
    store = mockStore(initialState);
  });

  /**
   * Test AuthScreen snapshot
   */
  it("Auth snapshot", () => {
    const snap = render(
      <MockedNavigator component={AuthScreen} store={store} />
    ).toJSON();
    expect(snap).toMatchSnapshot();
  });

  /**
   * Submit Button should be disable if username and password are empty
   */
  it("Login button enable", () => {
    const { getByTestId } = render(
      <MockedNavigator component={AuthScreen} store={store} />
    );

    const userNameText = getByTestId("UserName");
    const passwordText = getByTestId("Password");
    const submitButton = getByTestId("Submit");

    // because i use a custom button we can not access disable property so i use accessibilityHint instead
    expect(submitButton.props.accessibilityHint).toEqual("true");

    fireEvent.changeText(userNameText, "mostafa");
    fireEvent.changeText(passwordText, "123");

    expect(submitButton.props.accessibilityHint).toEqual("false");
  });
});

describe("Authentication action", () => {
  let initialState;
  let store;

  beforeEach(() => {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);

    // Initialize mockstore
    initialState = {
      auth: {
        authentication: {},
        token: undefined,
      },
    };
    store = mockStore(initialState);
  });

  beforeAll(() => {
    return mocker.enableMocking();
  });

  afterAll(() => {
    return mocker.disableMocking();
  });

  it("right credentials", async () => {
    const expectedActions = [
      {
        type: AuthActionTypes.AUTHENTICATION,
        status: AsyncActionStatus.REQUEST,
      },
      {
        type: AuthActionTypes.AUTHENTICATION,
        status: AsyncActionStatus.SUCCESS,
        data: mockAuthResponse,
      },
    ];

    return store.dispatch(authAsyncAction("mostafa", "123")).then(() => {
      // return of  actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it("wrong credentials", async () => {
    const expectedActions = [
      {
        type: AuthActionTypes.AUTHENTICATION,
        status: AsyncActionStatus.REQUEST,
      },
      {
        type: AuthActionTypes.AUTHENTICATION,
        status: AsyncActionStatus.FAILURE,
        error: Error("wrong credentials"),
      },
    ];

    return store.dispatch(authAsyncAction("dddd", "2222")).then(() => {
      // return of  actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
