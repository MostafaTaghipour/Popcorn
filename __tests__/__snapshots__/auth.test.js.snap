// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth render Auth snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaView
    onInsetsChange={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        pointerEvents="box-none"
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          closing={false}
          gestureVelocityImpact={0.3}
          importantForAccessibility="auto"
          onClose={[Function]}
          onGestureBegin={[Function]}
          onGestureCanceled={[Function]}
          onOpen={[Function]}
          onTransitionStart={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          transitionSpec={
            Object {
              "close": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
              "open": Object {
                "animation": "spring",
                "config": Object {
                  "damping": 500,
                  "mass": 3,
                  "overshootClamping": true,
                  "restDisplacementThreshold": 0.01,
                  "restSpeedThreshold": 0.01,
                  "stiffness": 1000,
                },
              },
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Object {
                "flex": 1,
                "marginTop": 64,
              }
            }
          >
            <View
              collapsable={false}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              style={
                Object {
                  "flex": 1,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <View
                pointerEvents="none"
                style={
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": -1,
                    },
                    "shadowOpacity": 0.3,
                    "shadowRadius": 5,
                    "top": 0,
                    "width": 3,
                  }
                }
              />
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "overflow": "hidden",
                    },
                    Array [
                      Object {
                        "backgroundColor": "rgb(242, 242, 242)",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                      }
                    }
                  >
                    <View
                      onLayout={[Function]}
                      style={
                        Array [
                          Object {
                            "flex": 1,
                          },
                          Object {
                            "paddingBottom": 0,
                          },
                        ]
                      }
                    >
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "height": 60,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <Text
                          style={
                            Array [
                              Object {
                                "color": "#000000",
                                "fontFamily": "System",
                                "fontWeight": "400",
                                "textAlign": "left",
                              },
                              Array [
                                Object {
                                  "color": "rgba(0, 0, 0, 0.87)",
                                  "fontFamily": "System",
                                  "fontWeight": "500",
                                  "textAlign": "left",
                                  "writingDirection": "ltr",
                                },
                                Array [
                                  Object {
                                    "fontSize": 20,
                                    "letterSpacing": 0.15,
                                    "lineHeight": 30,
                                    "marginVertical": 2,
                                  },
                                  undefined,
                                ],
                                Array [
                                  Object {
                                    "fontSize": 20,
                                    "letterSpacing": 0.15,
                                    "lineHeight": 30,
                                    "marginVertical": 2,
                                  },
                                  undefined,
                                ],
                              ],
                            ]
                          }
                        >
                          
                        </Text>
                        <Text />
                      </View>
                      <View
                        style={
                          Object {
                            "marginHorizontal": 16,
                            "marginVertical": 8,
                          }
                        }
                      >
                        <View>
                          <View
                            pointerEvents="none"
                            style={
                              Array [
                                Object {
                                  "bottom": 0,
                                  "left": 0,
                                  "position": "absolute",
                                  "right": 0,
                                  "top": 6,
                                },
                                Object {
                                  "backgroundColor": "#f6f6f6",
                                  "borderColor": "rgba(0, 0, 0, 0.54)",
                                  "borderRadius": 4,
                                  "borderWidth": 1,
                                },
                              ]
                            }
                          />
                          <View
                            style={
                              Object {
                                "minHeight": 64,
                                "paddingBottom": 0,
                                "paddingTop": 8,
                              }
                            }
                          >
                            <TextInput
                              allowFontScaling={true}
                              editable={true}
                              multiline={false}
                              onBlur={[Function]}
                              onChangeText={[Function]}
                              onFocus={[Function]}
                              placeholderTextColor="rgba(0, 0, 0, 0.54)"
                              rejectResponderTermination={true}
                              selectionColor="#6200ee"
                              style={
                                Array [
                                  Object {
                                    "flexGrow": 1,
                                    "margin": 0,
                                    "paddingHorizontal": 14,
                                    "textAlign": "left",
                                    "zIndex": 1,
                                  },
                                  Object {
                                    "height": 56,
                                  },
                                  Object {
                                    "paddingBottom": 0,
                                    "paddingTop": 0,
                                  },
                                  Object {
                                    "color": "#000000",
                                    "fontFamily": "System",
                                    "fontSize": 16,
                                    "fontWeight": undefined,
                                    "textAlignVertical": "center",
                                  },
                                ]
                              }
                              testID="UserName"
                              underlineColorAndroid="transparent"
                            />
                          </View>
                        </View>
                      </View>
                      <View
                        style={
                          Object {
                            "marginHorizontal": 16,
                            "marginVertical": 8,
                          }
                        }
                      >
                        <View
                          style={Object {}}
                        >
                          <View>
                            <View
                              pointerEvents="none"
                              style={
                                Array [
                                  Object {
                                    "bottom": 0,
                                    "left": 0,
                                    "position": "absolute",
                                    "right": 0,
                                    "top": 6,
                                  },
                                  Object {
                                    "backgroundColor": "#f6f6f6",
                                    "borderColor": "rgba(0, 0, 0, 0.54)",
                                    "borderRadius": 4,
                                    "borderWidth": 1,
                                  },
                                ]
                              }
                            />
                            <View
                              style={
                                Object {
                                  "minHeight": 64,
                                  "paddingBottom": 0,
                                  "paddingTop": 8,
                                }
                              }
                            >
                              <TextInput
                                allowFontScaling={true}
                                editable={true}
                                multiline={false}
                                onBlur={[Function]}
                                onChangeText={[Function]}
                                onFocus={[Function]}
                                placeholderTextColor="rgba(0, 0, 0, 0.54)"
                                rejectResponderTermination={true}
                                secureTextEntry={true}
                                selectionColor="#6200ee"
                                style={
                                  Array [
                                    Object {
                                      "flexGrow": 1,
                                      "margin": 0,
                                      "paddingHorizontal": 14,
                                      "textAlign": "left",
                                      "zIndex": 1,
                                    },
                                    Object {
                                      "height": 56,
                                    },
                                    Object {
                                      "paddingBottom": 0,
                                      "paddingTop": 0,
                                    },
                                    Object {
                                      "color": "#000000",
                                      "fontFamily": "System",
                                      "fontSize": 16,
                                      "fontWeight": undefined,
                                      "textAlignVertical": "center",
                                    },
                                  ]
                                }
                                testID="Password"
                                underlineColorAndroid="transparent"
                              />
                            </View>
                          </View>
                        </View>
                        <Text />
                      </View>
                      <View
                        style={
                          Object {
                            "alignItems": "center",
                            "alignSelf": "center",
                            "backgroundColor": "rgba(0, 0, 0, 0.12)",
                            "borderColor": "transparent",
                            "borderRadius": 4,
                            "borderStyle": "solid",
                            "borderWidth": 0,
                            "elevation": 0,
                            "height": 50,
                            "justifyContent": "center",
                            "marginTop": 32,
                            "minWidth": 64,
                            "width": 200,
                          }
                        }
                      >
                        <View
                          accessibilityRole="button"
                          accessibilityStates={
                            Array [
                              "disabled",
                            ]
                          }
                          accessible={true}
                          focusable={true}
                          isTVSelectable={true}
                          onClick={[Function]}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            Array [
                              Object {
                                "overflow": "hidden",
                              },
                              Object {
                                "borderRadius": 4,
                              },
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "flexDirection": "row",
                                  "justifyContent": "center",
                                },
                                undefined,
                              ]
                            }
                          >
                            <Text
                              numberOfLines={1}
                              style={
                                Array [
                                  Object {
                                    "color": "#000000",
                                    "fontFamily": "System",
                                    "fontWeight": "400",
                                    "textAlign": "left",
                                  },
                                  Array [
                                    Object {
                                      "letterSpacing": 1,
                                      "marginHorizontal": 16,
                                      "marginVertical": 9,
                                      "textAlign": "center",
                                    },
                                    undefined,
                                    Object {
                                      "textTransform": "uppercase",
                                    },
                                    Object {
                                      "color": "rgba(0, 0, 0, 0.32)",
                                      "fontFamily": "System",
                                      "fontWeight": "500",
                                    },
                                    Object {
                                      "fontFamily": "System",
                                      "fontWeight": "500",
                                    },
                                    undefined,
                                  ],
                                ]
                              }
                            >
                              <Text
                                accessibilityHint="true"
                                style={
                                  Array [
                                    Object {
                                      "color": "#000000",
                                      "fontFamily": "System",
                                      "fontWeight": "400",
                                      "textAlign": "left",
                                    },
                                    undefined,
                                  ]
                                }
                                testID="Submit"
                              >
                                
                              </Text>
                            </Text>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      pointerEvents="box-none"
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        onLayout={[Function]}
        pointerEvents="box-none"
        style={null}
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgb(255, 255, 255)",
                "borderBottomColor": "rgb(224, 224, 224)",
                "flex": 1,
                "shadowColor": "rgb(224, 224, 224)",
                "shadowOffset": Object {
                  "height": 0.5,
                  "width": 0,
                },
                "shadowOpacity": 0.85,
                "shadowRadius": 0,
              }
            }
          />
        </View>
        <View
          pointerEvents="box-none"
          style={
            Object {
              "height": 64,
              "maxHeight": undefined,
              "minHeight": undefined,
              "opacity": undefined,
              "transform": undefined,
            }
          }
        >
          <View
            pointerEvents="none"
            style={
              Object {
                "height": 20,
              }
            }
          />
          <View
            pointerEvents="box-none"
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "marginHorizontal": 16,
                  "opacity": 1,
                }
              }
            >
              <Text
                accessibilityRole="header"
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "rgb(28, 28, 30)",
                    "fontSize": 17,
                    "fontWeight": "600",
                  }
                }
              >
                MockedScreen
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
